lib/graphql/input_object_type.rb

```diff
@@ -51,6 +51,18 @@ def kind
     def validate_non_null_input(input, warden)
       result = GraphQL::Query::InputValidationResult.new
 
+      if input.is_a? Array
```

Comment:
Sorry to go quiet on this one, I've just been stumped by it! 

Yes, `respond_to?(:each)` isn't even _quite_ right. I can think of two options, like the ones you just described: 

- Check for `#each`, and assume that it will yield key-value pairs. (This assumption may be wrong, eg Array) 
- Check for `#each`, and inside the block, assert that two values were yielded. I don't think this is really possible: the second yield arg may be `nil`, in which case, we can't tell whether the user _provided_ nil, or whether nil is expressing "nothing was yielded"
- "Normalize" the input with something like `val_h = value.is_a?(Hash) ? value : value.to_h`. This could work, I wish it was more "duck type"-y, but I can't think of anything really _wrong_ with it. 

What do you think of those?